### PR TITLE
Implement the QANet model on SQuAD

### PR DESCRIPTION
**Pull request by yizhongw**

### CODE REVIEW

File: .gitignore
Comment (matt-gardner) [inline .gitignore]: I'd recommend against this rule, as it will also catch python files we add under a `data` module (or test fixtures, etc.).

In general, things that are specific to your setup should be in your own local `.gitignore` (either like [this](https://medium.com/@peter_graham/how-to-create-a-local-gitignore-1b19f083492b) or like [this](https://gist.github.com/subfuzion/db7f57fff2fb6998a16c)).  (I think this applies to all four of these rules.)

File: reading_comprehension/ema_trainer.py
Comment (matt-gardner) [inline reading_comprehension/ema_trainer.py]: The changes you're adding here are small and generally useful.  These should go in the main repo.  You don't need to worry about this right now, but at some point I'd recommend making a PR on the main repo that adds an option for exponential moving average to the trainer.  There are a few style things that need to be fixed when you make that PR:

- `self.ema` is not a good name - if I come across `self.ema` down below, at any line other than this one, I will have no idea what it means.  It's a bit annoying to have a long name that you call long method names on, but that's better than having to search through the file to figure out what the name means.  Something like `self.moving_average` would be fine.
- `ExponentialMovingAverage` should go into its own file in the main repo, as `trainer.py` is already quite large.

File: reading_comprehension/ema_trainer.py
Comment (matt-gardner) [inline reading_comprehension/ema_trainer.py]: I'd probably combine these two parameters into one - just have the default value for `exponential_moving_average_decay` be `None`, and that means the same as `use_exponential_moving_average=False`.

File: reading_comprehension/ema_trainer.py
Comment (matt-gardner) [inline reading_comprehension/ema_trainer.py]: Because this is our prototype code, not the main library, it's not as important to have super detailed docstrings, but this should at least say what the difference is between this and the base `Trainer` class.  It'd also be helpful to have comments below where you've added stuff saying what you added.  This is very nice in cases where you copy large chunks of code and modify them just a little bit.

File: reading_comprehension/ema_trainer.py
Comment (matt-gardner) [inline reading_comprehension/ema_trainer.py]: So in here you can say something like "Exactly the same as `Trainer._train_epoch` except for the addition of a call to `self.ema.apply()`.

File: reading_comprehension/ema_trainer.py
Comment (matt-gardner) [inline reading_comprehension/ema_trainer.py]: You're allowing for `self.ema` to be `None` in the constructor, but that will make this crash.

File: reading_comprehension/ema_trainer.py
Comment (matt-gardner) [inline reading_comprehension/ema_trainer.py]: Similarly here - say what's different from the main class.

File: reading_comprehension/ema_trainer.py
Comment (matt-gardner) [inline reading_comprehension/ema_trainer.py]: Type hints on all parameters in this class.

File: reading_comprehension/squad_reader.py
Comment (matt-gardner) [inline reading_comprehension/squad_reader.py:18]: What did you have to change here?

File: reading_comprehension/utils.py
Comment (matt-gardner) [inline reading_comprehension/utils.py]: Modules shouldn't be in a `utils.py` file, it makes them much harder to find.  I'd add a `modules/` directory, and have each module have its own file in there.

File: reading_comprehension/utils.py
Comment (matt-gardner) [inline reading_comprehension/utils.py]: Avoid things like `idx` - having your code be immediately readable (without having to figure out what your abbreviation means) is more important than saving a few characters.  See https://google.github.io/styleguide/cppguide.html#General_Naming_Rules.

Comment (matt-gardner) [reply reading_comprehension/utils.py]: Also, add type hints here.

File: reading_comprehension/utils.py
Comment (matt-gardner) [inline reading_comprehension/utils.py]: I'm surprised this works, dropping out the entire layer.  This was much more aggressive than I expected.

File: reading_comprehension/utils.py
Comment (matt-gardner) [inline reading_comprehension/utils.py]: Type hints on all of this.

File: reading_comprehension/utils.py
Comment (matt-gardner) [inline reading_comprehension/utils.py]: A note here on how this is different from our normal `masked_softmax` would be good.

File: tests/test_qanet.py
Comment (matt-gardner) [inline tests/test_qanet.py]: The fixtures are inside the `tests` directory in the main allennlp repo because we need them to run tests when installed with pip.  I think it's easier to keep them outside.  So I'd recommend just having a top-level `fixtures` directory, instead of doing it like this.

File: reading_comprehension/utils.py
Comment (matt-gardner) [inline reading_comprehension/utils.py]: More helpful would be something like: `f"We currently only handle 1 and 2 dimensional convolutions here.  You gave {dim}."`

**Review by matt-gardner — COMMENTED** [review]
I haven't looked at the model or the encoder yet, but I have to go for now.  I'll try to look at them later tonight.  What I've looked at so far looks pretty good!

File: reading_comprehension/utils.py
Comment (yizhongw) [reply reading_comprehension/utils.py]: Yes, I guess this works probably because it's applied to the residual network. The next layer makes prediction based on all previous layers, so the dropping of one layer might not make much difference.

File: reading_comprehension/modules/depth_sep_conv.py
Comment (matt-gardner) [inline reading_comprehension/modules/depth_sep_conv.py]: Even in the filename, I'd avoid abbreviations.  `conv` is probably fine; I'd recommend `depthwise_separable_conv.py`.  It's ok to have long names - better that than making the reader guess what you meant.

File: reading_comprehension/modules/depth_sep_conv.py
Comment (matt-gardner) [inline reading_comprehension/modules/depth_sep_conv.py]: Where is this needed?  You don't even have a logger in here.

File: reading_comprehension/modules/layer_dropout.py
Comment (matt-gardner) [inline reading_comprehension/modules/layer_dropout.py]: f-strings are nicer: `f"but got {undecayed_dropout_prob}"`

File: reading_comprehension/modules/layer_dropout.py
Comment (matt-gardner) [inline reading_comprehension/modules/layer_dropout.py]: Missing period.

File: reading_comprehension/modules/layer_dropout.py
Comment (matt-gardner) [inline reading_comprehension/modules/layer_dropout.py]: f-strings here, too.

File: reading_comprehension/modules/layer_dropout.py
Comment (matt-gardner) [inline reading_comprehension/modules/layer_dropout.py]: "The output tensor", yes?

File: reading_comprehension/modules/layer_dropout.py
Comment (matt-gardner) [inline reading_comprehension/modules/layer_dropout.py:48]: I think you need a blank line before this section for sphinx to render it properly (only matters once we merge this back upstream).

File: reading_comprehension/qanet.py
Comment (matt-gardner) [inline reading_comprehension/qanet.py]: Nit: we try to use [Google's definition of camel case](https://google.github.io/styleguide/javaguide.html#s5.3-camel-case), where this would be `QaNet`.  It makes it easier to read, so you don't have to guess where one acronym ends and another starts (imagine, e.g., `QALSTMNet` - much easier to parse if it's `QaLstmNet`).

File: reading_comprehension/qanet.py
Comment (matt-gardner) [inline reading_comprehension/qanet.py:90]: Is it worth having any consistency checks on the dimensions of your inputs?  I'm thinking of something like this: https://github.com/allenai/allennlp/blob/aa1b774ed8de31ec04bebf9f054200bc2507e0c5/allennlp/models/reading_comprehension/bidaf.py#L96-L103

File: reading_comprehension/qanet.py
Comment (matt-gardner) [inline reading_comprehension/qanet.py:190]: Have you tried removing the masks from these encoder operations?  Does that change anything?

File: reading_comprehension/qanet.py
Comment (matt-gardner) [inline reading_comprehension/qanet.py:206]: Looks like we should pull `get_best_span` out into a `reading_comprehension.utils` module.  But that can be done later.

File: reading_comprehension/qanet_encoder.py
Comment (matt-gardner) [inline reading_comprehension/qanet_encoder.py]: `QaNetEncoderBlock`

File: reading_comprehension/qanet_encoder.py
Comment (matt-gardner) [inline reading_comprehension/qanet_encoder.py]: This should just be `super().__init__()`.

File: reading_comprehension/qanet_encoder.py
Comment (matt-gardner) [inline reading_comprehension/qanet_encoder.py]: `super().__init__()`

File: tests/test_qanet.py
Comment (matt-gardner) [inline tests/test_qanet.py]: `super().setUp()`

File: reading_comprehension/ema_trainer.py
Comment (matt-gardner) [inline reading_comprehension/ema_trainer.py]: `super().__init__(...)`

File: reading_comprehension/modules/depth_sep_conv.py
Comment (matt-gardner) [inline reading_comprehension/modules/depth_sep_conv.py]: `super().__init__()`

File: reading_comprehension/modules/layer_dropout.py
Comment (matt-gardner) [inline reading_comprehension/modules/layer_dropout.py]: `super().__init__()`

File: reading_comprehension/qanet.py
Comment (matt-gardner) [inline reading_comprehension/qanet.py]: `super().__init__()`

File: reading_comprehension/qanet_encoder.py
Comment (matt-gardner) [inline reading_comprehension/qanet_encoder.py]: I'd recommend using `check_dimensions_match` here, just to give a more informative error message.

File: reading_comprehension/qanet_encoder.py
Comment (matt-gardner) [inline reading_comprehension/qanet_encoder.py]: I think there's a `ModuleList` that you can use to make this a little cleaner; not a big deal, though.

File: reading_comprehension/qanet_encoder.py
Comment (matt-gardner) [inline reading_comprehension/qanet_encoder.py]: `count`, not `cnt` - avoid abbreviations.  I'd similarly recommend `attention` instead of `attn` and `feedforward` or `feed_forward` instead of `ffn`, but I don't feel quite as strongly about those.

File: reading_comprehension/qanet_encoder.py
Comment (matt-gardner) [inline reading_comprehension/qanet_encoder.py]: `StackedQaNetEncoder`

File: reading_comprehension/qanet_encoder.py
Comment (matt-gardner) [inline reading_comprehension/qanet_encoder.py]: `index`, not `idx`.

File: reading_comprehension/qanet_encoder.py
Comment (matt-gardner) [inline reading_comprehension/qanet_encoder.py:212]: You could do:

```python
if input_dim != hidden_dim:
    self._input_projection_layer = ...
else:
    self._input_projection_layer = lambda x: x
```

Then, in `forward` below, you don't need to have an `if` checking for whether `input_projection_layer is None`.

File: reading_comprehension/qanet_encoder.py
Comment (matt-gardner) [inline reading_comprehension/qanet_encoder.py]: I'm pretty sure `self.dropout` is not defined here.  Is this module tested?

File: reading_comprehension/qanet_encoder.py
Comment (matt-gardner) [inline reading_comprehension/qanet_encoder.py:257]: These modifications should just be submitted upstream, I think.  It's fine to have them in here for now, though, until it gets merged upstream.

File: reading_comprehension/qanet_encoder.py
Comment (matt-gardner) [reply reading_comprehension/qanet_encoder.py]: Actually, probably just `QaNetEncoder` - no need to have the `Stacked` in the name, as the `Block` in the other name already implies that.

**Review by matt-gardner — APPROVED** [review]
I had a bunch of comments, but they're pretty much all stylistic.  Looks great!  Feel free to merge when you're happy with it (and if you don't have permission to merge yet, let me know and I'll fix that).

File: reading_comprehension/qanet.py
Comment (yizhongw) [reply reading_comprehension/qanet.py:190]: Yes, that mask works in the self-attention part.

**Comment by matt-gardner** [comment]
Looks like CI is still failing - definitely don't merge until that passes, because if it's broken then no other PR will pass, even if it doesn't introduce anything new that breaks things.

**Comment by yizhongw** [comment]
Sorry, I missed some changes in my last commit. So I committed again with `--amend`.

**Pull request merged**
